array_analyzer/extract/metadata.py

```diff
@@ -0,0 +1,221 @@
+        c.FIDUCIAL_ARRAY = txt_parser.populate_array_fiduc(self.fiducials_array, c.fiducials)
+
+    def _create_antigen_array(self):
+        """
+        Creates an empty ndarray of strings, whose rows and columns match the printed array's rows/cols
+            Assigns the corresponding "Antigen" based on metadata
+            "Antigens" are descriptive values of the antigen at each spot location.
+
+        This is the only array creator that requires one of the above arrays (SPOT_ID_ARRAY, .xml meta ONLY)
+            multiple "spot-ids" can contain the same "antigen".  Thus it's required to pass the "SPOT_ID_ARRAY"
+        :return:
+        """
+        self.antigen_array = np.empty(shape=(c.params['rows'], c.params['columns']), dtype='U100')
+        if c.METADATA_EXTENSION == 'xml':
+            if c.SPOT_ID_ARRAY.size == 0:
+                raise AttributeError("attempting to create antigen array before SPOT_ID_ARRAY is assigned")
+            c.ANTIGEN_ARRAY = txt_parser.populate_array_antigen_xml(self.antigen_array, c.SPOT_ID_ARRAY, c.replicates)
+        elif c.METADATA_EXTENSION == 'csv' or c.METADATA_EXTENSION == 'xlsx':
+            c.ANTIGEN_ARRAY = txt_parser.populate_array_antigen(self.antigen_array, c.replicates)
+
+    def _calculate_fiduc_coords(self):
+        """
+        Calculate and set the fiducial coordinates like:
+            FIDUCIALS = [(0, 0), (0, 1), (0, 5), (7, 0), (7, 5)]
+            fiducial coordinates are labeled as "Reference, Diagnostic"
+        :return:
+        """
+        x, y = np.where(c.FIDUCIAL_ARRAY == 'Reference, Diagnostic')
+        if x.size == 0 or y.size == 0:
+            x, y = np.where(c.FIDUCIAL_ARRAY == 'Fiducial')
+        c.FIDUCIALS = list(zip(x, y))
+
+    def _calculate_fiduc_idx(self):
+        """
+        Calculate fiducial index like
+            FIDUCIALS_IDX = [0, 5, 6, 30, 35]\
+
+
+            FIDUCIALS_IDX = [0, 7, 8, 40, 47] for 8 columns
+        :return:
+        """
+        c.FIDUCIALS_IDX = list(np.where(c.FIDUCIAL_ARRAY.flatten() == 'Reference, Diagnostic')[0])
+        if len(c.FIDUCIALS_IDX) == 0:
+            c.FIDUCIALS_IDX = list(np.where(c.FIDUCIAL_ARRAY.flatten() == 'Fiducial')[0])
+
+    def _calc_spot_dist(self):
+        """
+        Calculate distance between spots in both pixels and microns
+        :return:
+        """
+        v_pitch_mm = c.params['v_pitch']
+        h_pitch_mm = c.params['h_pitch']
+        pix_size = c.params['pixel_size']
+
+        # assuming similar v_pitch and h_pitch, average to get the SPOT_DIST
+        v_pitch_pix = v_pitch_mm/pix_size
+        h_pitch_pix = h_pitch_mm/pix_size
+        c.SPOT_DIST_PIX = np.mean([v_pitch_pix, h_pitch_pix]).astype('uint8')
+
+        # convert the SPOT_DIST to microns, 0 - 255
+        c.SPOT_DIST_UM = np.mean([v_pitch_mm * 1000, h_pitch_mm * 1000]).astype('uint8')
+
+    # set filesaving run_path
+    def _set_run_path(self, output_folder):
```

Comment:
Just FIY I added a similar function in find_images, but we can just remove whichever you prefer when merging branches.

Reply:
I think i'm going to keep the version in MetaData.  Though, I can go eiter way if you have a preference.